ssz/src/main/java/tech/pegasys/artemis/util/backing/view/AbstractCompositeViewRead.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.backing.view;
+
+import tech.pegasys.artemis.util.backing.CompositeViewRead;
+import tech.pegasys.artemis.util.backing.ViewRead;
+import tech.pegasys.artemis.util.backing.cache.ArrayCache;
+import tech.pegasys.artemis.util.backing.cache.IntCache;
+import tech.pegasys.artemis.util.backing.tree.TreeNode;
+import tech.pegasys.artemis.util.backing.type.CompositeViewType;
+
+/**
+ * Base backing view class for immutable composite views (lists, vectors, containers)
+ *
+ * <p>It caches it's child view instances so that if the underlying tree nodes are not changed (in
+ * the corresponding mutable classes) the instances are not recreated from tree nodes on later
+ * access.
+ *
+ * <p>Thought internally this class has a mutable cache it may be thought of as immutable instance
+ * and used safely across threads
+ *
+ * @param <ChildType> the type of children. For heterogeneous composites (like container) this type
+ *     would be just generic {@link ViewRead}
+ */
+public abstract class AbstractCompositeViewRead<ChildType extends ViewRead>
+    implements CompositeViewRead<ChildType> {
+
+  private IntCache<ChildType> childrenViewCache;
+  private final int sizeCache;
+  private final CompositeViewType type;
+  private final TreeNode backingNode;
+
+  /** Creates an instance from a type and a backing node */
+  protected AbstractCompositeViewRead(CompositeViewType type, TreeNode backingNode) {
```

Comment:
Should we be passing a `Supplier` for the backingNode so that we only produce the backing tree if we need it??
```suggestion
  protected AbstractCompositeViewRead(CompositeViewType type, Supplier<TreeNode> backingNode) {
```

Reply:
What use cases do you mean here? 
Like when you create an immutable container with field values?  Like here: 
https://github.com/PegaSysEng/teku/blob/a01633de6ded2fb9d7c464dd8bd73a9cbb1f46d7/ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/state/Fork.java#L59-L65

Or something different?

Reply:
Yeah - for example [here](https://github.com/Nashatyrev/artemis/blob/af118aa35f2e3b92e303db06f717b66595fb75f8/ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/GenesisGenerator.java#L103-L110) we create a few intermediate versions of a `Validator` and we build a tree for each one even though we shouldn't need it for the intermediate ones.

Reply:
Yeah I see what you mean. That indeed looks sub optimal. 

However I would refrain from introducing this lambda for all views since it would most likely have negative overall performance impact, since normally a `View` either should already has backing tree node or would need it later.  
So the only acceptable solution I see is to refactor `AbstractImmutableContainer` for such cases but I would not invest time in it as long as this is not a bottleneck now (i.e. avoid preliminary optimization). 

The case you posted I optimized in the this way: https://github.com/PegaSysEng/teku/pull/1465/commits/1fe4589e4b05242d47f5ae5fccfd561250552f9b It is more verbose but shouldn't have overhead anymore : the Validator tree (created internally) should be used later.